magenta/models/polyphonic_rnn/polyphonic_rnn_lib.py

```diff
@@ -0,0 +1,1377 @@
+  return tf.reverse(ti, am)
+
+
+def ni(t, ind, axis=0):
+  # Negative single index helper
+  ndims = len(shape(t))
+  im1 = -1 + abs(ind)
+  i = [[None, None]] * ndims
+  i[axis] = [im1, im1 + 1]
+  am = [False] * ndims
+  am[axis] = True
+  sl = [slice(*ii) for ii in i]
+  ti = tf.reverse(t, am)[sl]
+  return ti[0, :, :]
+
+
+def scan(fn, sequences, outputs_info):
+  # for some reason TF step needs initializer passed as first argument?
+  # a tiny wrapper to tf.scan to make my life easier
+  # closer to theano scan, allows for step functions with multiple arguments
+  # may eventually have kwargs which match theano
+  for i in range(len(sequences)):
+    # Try to accomodate for masks...
+    seq = sequences[i]
+    nd = ndim(seq)
+    if nd == 3:
+      pass
+    elif nd < 3:
+      sequences[i] = tf.expand_dims(sequences[i], nd)
+    else:
+      raise ValueError('Ndim too different to correct')
+
+  def check(l):
+    shapes = [shape(s) for s in l]
+    # for now assume -1, can add axis argument later
+    # check shapes match for concatenation
+    compat = [ls for ls in shapes if ls[:-1] == shapes[0][:-1]]
+    if len(compat) != len(shapes):
+      raise ValueError('Tensors *must* be the same dim for now')
+
+  check(sequences)
+  check(outputs_info)
+
+  seqs_shapes = [shape(s) for s in sequences]
+  nd = len(seqs_shapes[0])
+  seq_pack = tf.concat(nd - 1, sequences)
+  outs_shapes = [shape(o) for o in outputs_info]
+  nd = len(outs_shapes[0])
+  init_pack = tf.concat(nd - 1, outputs_info)
+
+  assert len(shape(seq_pack)) == 3
+  assert len(shape(init_pack)) == 2
+
+  def s_e(shps):
+    starts = []
+    ends = []
+    prev_shp = 0
+    for _, shp in enumerate(shps):
+      start = prev_shp
+      end = start + shp[-1]
+      starts.append(start)
+      ends.append(end)
+      prev_shp = end
+    return starts, ends
+
+  # TF puts the initializer in front?
+  def fnwrap(initializer, elems):
+    starts, ends = s_e(seqs_shapes)
+    sliced_elems = [elems[:, start:end] for start, end in zip(starts, ends)]
+    starts, ends = s_e(outs_shapes)
+    sliced_inits = [initializer[:, start:end]
+                    for start, end in zip(starts, ends)]
+    t = []
+    t.extend(sliced_elems)
+    t.extend(sliced_inits)
+    # elems first then inits
+    outs = fn(*t)
+    nd = len(outs_shapes[0])
+    outs_pack = tf.concat(nd - 1, outs)
+    return outs_pack
+
+  r = tf.scan(fnwrap, seq_pack, initializer=init_pack)
+
+  if len(outs_shapes) > 1:
+    starts, ends = s_e(outs_shapes)
+    o = [r[:, :, start:end] for start, end in zip(starts, ends)]
+    return o
+  else:
+    return r
+
+##
+# end metautils
+##
+
+##
+# begin datasets
+##
+
+
+def duration_and_pitch_to_midi(filename, durations, pitches, prime_until=0):
+  # durations and pitches should both be 2D
+  # [time_steps, n_notes]
+
+  sequence = music_pb2.NoteSequence()
+
+  # Hardcode for now, eventually randomize?
+  # or predict...
+  sequence.ticks_per_quarter = mm.STANDARD_PPQ
```

Comment:
Yeah, I suspect I'll have to make a lot more things user-controlled when I start adding the sequence generator interface.